src/target/riscv/riscv.c

```diff
@@ -2926,6 +2931,13 @@ int riscv_openocd_poll(struct target *target)
 		}
 	}
 
+	foreach_smp_target(entry, targets) {
+		struct target *t = entry->target;
+		struct riscv_info *info = riscv_info(t);
+		if (info->tick && info->tick(t) != ERROR_OK)
+			return ERROR_FAIL;
+	}
+
```

Comment:
What case is this loop supposed to cover? My understanding is that we iterate over the individual harts already above (line 2871) and the dcsr.ebreak* manipulation should be handled inside riscv_poll_hart() (called on line 2883).

Is this loop to cover the `unavailable --> running` transition? If so, then it appears cleaner to put the logic into the newly added `handle_running` callback (and to pass also the previous state to the handler as discussed above).

Reply:
The reason for this loop is to call tick() for every hart. What happens in tick() is opaque to this layer. The reason it's outside the previous loop is that at this point the state of every hart has settled, so any side effects happening in tick() won't affect the delicate poll() code.

Reply:
Ok, understood, thank you for the explanation. 

If you like, you can put this as a short comment on top of this "foreach".

Reply:
Done.